.github/workflows/tf-apply-hash.yml

```diff
@@ -0,0 +1,129 @@
+    runs-on: ubuntu-latest
+    outputs:
+      matrix: ${{ steps.set-matrix.outputs.matrix }}
+    steps:
+      - name: Set matrix
+        id: set-matrix
+        env:
+          ENV: ${{ github.event.inputs.env }}
+        run: |
+          # If no env is specified, run both dev and prod
+          if [ "$ENV" == "prod" ]; then
+            echo "matrix=[\"$ENV\"]" >> $GITHUB_OUTPUT
+          else
+            echo "matrix=[\"prod\"]" >> $GITHUB_OUTPUT
+          fi
+
+  terraform:
+    runs-on: ubuntu-latest
+    permissions:
+      id-token: write
+      contents: read
+    env:
+      working-directory: infra/terraform/hash
+      TF_VAR_in_ci: true
+      AWS_PAGER:
+    needs:
+      - matrix-eval
+    strategy:
+      matrix:
+        env: ${{ fromJson(needs.matrix-eval.outputs.matrix) }}
+    steps:
+      - uses: actions/checkout@v3
+      - uses: hashicorp/setup-terraform@v2
+
+      - name: Prepare SSH key
+        working-directory: ${{ env.working-directory }}
+        run: |
+          echo "${{ secrets.BASTION_SSH_KEY }}" > h-hash-prod-usea1-bastionkey.pem
+
+      - name: Authenticate Vault
+        id: secrets
+        uses: hashicorp/vault-action@v2
+        with:
+          exportToken: true
+          url: ${{ env.VAULT_ADDR }}
+          method: jwt
+          role: ${{ matrix.env }}
+          # Even though it could look like separate calls to fetch the secrets
+          # the responses here are cached, so we're only issuing a single set of credentials
+          secrets: |
+            aws/creds/${{ matrix.env }}-deploy access_key | AWS_ACCESS_KEY_ID ;
+            aws/creds/${{ matrix.env }}-deploy secret_key | AWS_SECRET_ACCESS_KEY ;
+            aws/creds/${{ matrix.env }}-deploy security_token | AWS_SESSION_TOKEN
+
+      - name: Configure AWS Credentials
+        uses: aws-actions/configure-aws-credentials@v2
+        with:
+          aws-access-key-id: ${{ steps.secrets.outputs.AWS_ACCESS_KEY_ID }}
+          aws-secret-access-key: ${{ steps.secrets.outputs.AWS_SECRET_ACCESS_KEY }}
+          aws-session-token: ${{ steps.secrets.outputs.AWS_SESSION_TOKEN }}
+          aws-region: us-east-1
+
+      - name: "[${{ matrix.env }}] Terraform Init"
+        id: init
+        env:
+          TF_WORKSPACE: ${{ matrix.env }}
+        working-directory: ${{ env.working-directory }}
+        run: terraform init -no-color
+
+      - name: Terraform Validate
+        id: validate
+        run: terraform validate -no-color
+
+      - name: "[${{ matrix.env }}] Terraform Apply"
+        id: apply
+        env:
+          TF_WORKSPACE: ${{ matrix.env }}
+        working-directory: ${{ env.working-directory }}
+        run: terraform apply -auto-approve -no-color -var-file=${{ matrix.env }}-usea1.tfvars -refresh=false
+
+      - uses: actions/github-script@v6
+        id: set-result
+        with:
+          script: |
+            const apply = `${{ steps.apply.outputs.stdout }}`.replaceAll(/^~/g, "!");
```

Comment:
Does GitHub CI have the concept of `stdout` and `stderr`? I wonder if `stderr` output would be omitted.

Reply:
There is a small risk of a data race condition and the two outputs will be mixed up, generally, this is fine I think. In the (probably unlikely case) this will happen the GH Action output will still have everything in the correct order.

Reply:
Addressed in bb59888e13bbcb5d5536c9768b144cdc9b073b10